syntax = "proto3";

package core_management;

// A message representing a function
message Function {
    string name = 1;
    string code = 2;
    string runtime_env = 3;
    string version = 4;
}

// API 1 : addFunctionToLibrary
// Request message for adding a function
message AddFunctionRequest {
    string function_name = 1;
    string code = 2;
    string runtime_env = 3;
    string version = 4;
}

// Response message for adding a function
message AddFunctionResponse {
    bool success = 1;
    string message = 2; // Make sure this line exists
}

// API 4: ListFunctions
// Request message for listing all functions (no additional input needed)
message ListFunctionsRequest {}

// Response message for listing functions
message ListFunctionsResponse {
    repeated Function functions = 1;
}

// API 5: GetFunctionDetails
// Request message for getting function details
message GetFunctionDetailsRequest {
    string function_name = 1;
}

// Response message for getting function details
message GetFunctionDetailsResponse {
    Function function = 1;
}
    
// Request message for publish/unpublish a function 
message PublishFunctionRequest {
    string function_name = 1;
}

// Response message for publish/unpublish a function
message PublishFunctionResponse {
    bool success = 1;
    string message = 2;
}

// API 10
message ArchiveFunctionRequest {
    string function_name = 1;
}

// Response message for adding a function
message ArchiveFunctionResponse {
    bool success = 1;
    string message = 2;
}

// Service definition
service CoreManagement {
    rpc AddFunction(AddFunctionRequest) returns (AddFunctionResponse);
    rpc PublishFunction(PublishFunctionRequest) returns (PublishFunctionResponse);
    rpc UnpublishFunction(PublishFunctionRequest) returns (PublishFunctionResponse);
    rpc ArchiveFunction(ArchiveFunctionRequest) returns (ArchiveFunctionResponse);
    rpc ListFunctions(ListFunctionsRequest) returns (ListFunctionsResponse);
    rpc GetFunctionDetails(GetFunctionDetailsRequest) returns (GetFunctionDetailsResponse);
}
